### ISSUE TITLE
External packages: Unable to use package built with different compiler

### ISSUE THREAD

**Issue by adamjstewart** [spack/spack#3166]
I'm trying out Spack's external package support for the first time and finding it to be very... broken.

Ok, so here's the situation. I want Intel to be my default compiler and MKL to be my default BLAS/LAPACK. So I built `intel-parallel-studio` with my system compiler through Spack. Then I created a `packages.yaml` like:
```yaml
packages:
  all:
    compiler: [intel]
    providers:
      blas: [intel-parallel-studio]
      lapack: [intel-parallel-studio]
```
Of course, the problem now becomes that as soon as a newer version of `intel-parallel-studio` is added, or as soon as Spack's hashing algorithm changes, or if I change compilers, it will try to rebuild `intel-parallel-studio`. To prevent this, I added it as an external package. Now we have:
```yaml
packages:
  all:
    compiler: [intel]
    providers:
      blas: [intel-parallel-studio]
      lapack: [intel-parallel-studio]
  intel-parallel-studio:
    paths:
      intel-parallel-studio@professional.2017.1%gcc@4.4.7+all+daal~ilp64+ipp+mkl~mpi~newdtags+openmp+rpath+shared+tools: /soft/spack-0.10.0/opt/spack/linux-centos6-x86_64/gcc-4.4.7/intel-parallel-studio-professional.2017.1-zaq5tbp664b2ye2gdx3ry35t3wd4rjn4
    buildable: False
```
Now, I'm seeing:
```
$ spack spec arpack
arpack@96%gcc@4.4.7 arch=linux-centos6-x86_64 
    ^intel-parallel-studio@professional.2017.1%gcc@4.4.7+all+daal~ilp64+ipp+mkl~mpi~newdtags+openmp+rpath+shared+tools arch=linux-centos6-x86_64 
```
Hmm, that's not `%intel`. The compiler for the dependency seems to be overriding the default compiler. Let's try explicitly specifying `%intel` on the command line:
```
$ spack spec arpack%intel
arpack@96%intel@17.0.1 cflags="-axCORE-AVX2,AVX" cxxflags="-axCORE-AVX2,AVX" fflags="-axCORE-AVX2,AVX"  arch=linux-centos6-x86_64 
    ^intel-mkl@2017.0.098%intel@17.0.1 cflags="-axCORE-AVX2,AVX" cxxflags="-axCORE-AVX2,AVX" fflags="-axCORE-AVX2,AVX" ~ilp64+openmp+shared arch=linux-centos6-x86_64 
```
Cool, now it isn't using `intel-parallel-studio` anymore. I have to use:
```
$ spack spec arpack%intel ^intel-parallel-studio
```
to get what I really want.

**Comment by adamjstewart** [comment]
P.S. I would love to be able to use Intel MKL when `%intel` and use OpenBLAS when not `%intel`, but I can't think of a way to code that in `packages.yaml`.

**Comment by davydden** [comment]
> as soon as a newer version of intel-parallel-studio is added... To prevent this, I added it as an external package

I usually _ping_ installation by specify versions/variants i.e. 
```
  intel-mkl:
   version: [11.3.3.210]
   paths:
     mkl@11.3.3.210: /opt/intel/mkl
   buildable: False
```

>  I would love to be able to use Intel MKL when %intel and use OpenBLAS when not %intel, but I can't think of a way to code that in packages.yaml.

that would be indeed nice to have!


p.s. I also don't understand why conretizer did not choose `intel` compiler with `inntel-parallel-studio` blas/lapack from the beginning.

**Comment by adamjstewart** [comment]
@davydden Thanks for the suggestion. By removing the compiler from the spec, I can workaround the concretization bugs and get it to always use `intel-parallel-studio`.

**Comment by citibeth** [comment]
I must be missing some key information here; why is rebuilding `intel-parallel-studio` a problem?  Why was it not built with the Intel compiler?

**Comment by adamjstewart** [comment]
> why is rebuilding `intel-parallel-studio` a problem? 

1. Time
2. Space
3. I might not have a tarball on hand and it can't automatically download it, leaving me dead in the water every time a new version is added
4. In terms of Lmod, you should be able to lock down module files based on the compiler you used to build them. I don't want duplicate Intel compiler installations.

> Why was it not built with the Intel compiler?

I like to build all of my compilers with the system compiler.

**Comment by weijianwen** [comment]
Thank you @adamjstewart .  After long time of struggling, your settings finanally make my externally-installed Intel MPI work. This may also help guys @cnelson3  and @pramodk in  #3019 .

**Comment by alalazo** [comment]
For the record, encountered the same problem when building for two different OSes:
```yaml
packages:

  ####
  # Common libraries that we delegate to the OS
  ####

  openssl:
    buildable: False
    paths:
      openssl@system%gcc@4.4.7: /usr/lib64
      openssl@system%gcc@4.8.5: /usr/lib64
      openssl@system%gcc@5.4.0: /usr/lib64
      openssl@system%gcc@6.3.0: /usr/lib64
      openssl@system%intel@17.0.2: /usr/lib64
```
gives on a `rhel7` after #4034:
```console
$ spack compiler list
==> Available compilers
-- gcc rhel6-x86_E5v1_IntelIB -----------------------------------
gcc@6.3.0  gcc@5.4.0  gcc@4.4.7

-- gcc rhel6-x86_E5v2 -------------------------------------------
gcc@6.3.0  gcc@5.4.0  gcc@4.4.7

-- gcc rhel6-x86_E5v2_IntelIB -----------------------------------
gcc@6.3.0  gcc@5.4.0  gcc@4.4.7

-- gcc rhel6-x86_E5v2_Mellanox_GPU ------------------------------
gcc@6.3.0  gcc@5.4.0  gcc@4.4.7

-- gcc rhel6-x86_E5v3_IntelIB -----------------------------------
gcc@6.3.0  gcc@5.4.0  gcc@4.4.7

-- gcc rhel7-x86_64 ---------------------------------------------
gcc@4.8.5

-- gcc rhel7-x86_E5v4_Mellanox ----------------------------------
gcc@4.8.5

$ spack spec openssl
Input spec
--------------------------------
openssl

Normalized
--------------------------------
openssl
    ^zlib

Concretized
--------------------------------
==> Error: No compilers with spec gcc@4.4.7 found for operating system rhel7 and target x86_E5v4_Mellanox.
Run 'spack compiler find' to add compilers.

$ spack spec openssl%gcc@4.8.5
Input spec
--------------------------------
openssl%gcc@4.8.5

Normalized
--------------------------------
openssl%gcc@4.8.5
    ^zlib

Concretized
--------------------------------
openssl@system%gcc@4.8.5 arch=linux-rhel7-x86_E5v4_Mellanox 

$ spack spec cmake%gcc@4.8.5
Input spec
--------------------------------
cmake%gcc@4.8.5

Normalized
--------------------------------
cmake%gcc@4.8.5

Concretized
--------------------------------
==> Error: No compilers with spec gcc@4.4.7 found for operating system rhel7 and target x86_E5v4_Mellanox.
Run 'spack compiler find' to add compilers.
```

**Comment by becker33** [comment]
@adamjstewart I believe this is resolved by #14934. Closing it as resolved, open it if I'm incorrect.
